eras/conway/impl/test/Test/Cardano/Ledger/Conway/RatifySpec.hs

```diff
@@ -113,3 +156,107 @@ drepsPropNoStake =
+  , stakeNoConfidence :: Coin
+  , stakeNotVoted :: Coin
+  }
+  deriving (Show)
+
+data Ratios = Ratios
+  { yes :: Rational
+  , no :: Rational
+  , abstain :: Rational
+  , alwaysAbstain :: Rational
+  , noConfidence :: Rational
+  }
+  deriving (Show)
+
+-- Prepare the stake distribution and votes according to the given ratios.
+drepsTestData ::
+  forall era.
+  Era era =>
+  Ratios ->
+  Gen (DRepTestData era)
+drepsTestData Ratios {yes, no, abstain, alwaysAbstain, noConfidence} = do
+  let inDreps = listOf (DRepCredential <$> (arbitrary @(Credential 'DRepRole (EraCrypto era))))
+  -- dreps <- resize 10 inDreps
+  dreps <- inDreps
+
+  let drepSize = length dreps
+      alwaysAbstainPct :: Word64 = pct alwaysAbstain
+      noConfidencePct :: Word64 = pct noConfidence
+      distr =
+        Map.alter
+          (\case _ -> Just (CompactCoin noConfidencePct))
+          DRepAlwaysNoConfidence
+          . Map.alter
+            (\case _ -> Just (CompactCoin alwaysAbstainPct))
+            DRepAlwaysAbstain
+          $ Map.fromList [(drep, CompactCoin 1) | drep <- dreps]
+      (drepsYes, drepsNo, drepsAbstain, rest) = splitByPct yes no abstain dreps
+      notVotedStake = length rest
+      votes =
+        Map.union
+          (Map.fromList [(cred, VoteYes) | DRepCredential cred <- drepsYes])
+          $ Map.union
+            (Map.fromList [(cred, VoteNo) | DRepCredential cred <- drepsNo])
+            (Map.fromList [(cred, Abstain) | DRepCredential cred <- drepsAbstain])
+      pct :: Integral a => Rational -> a
+      pct r = ceiling (r * fromIntegral drepSize)
+  pure
+    DRepTestData
+      { distr = distr
+      , votes = votes
+      , totalStake = fromCompact (fold distr)
+      , stakeYes = Coin (fromIntegral (length drepsYes))
+      , stakeNo = Coin (fromIntegral (length drepsNo))
+      , stakeAbstain = Coin (fromIntegral (length drepsAbstain))
+      , stakeAlwaysAbstain = Coin (fromIntegral alwaysAbstainPct)
+      , stakeNoConfidence = Coin (fromIntegral noConfidencePct)
+      , stakeNotVoted = Coin (fromIntegral notVotedStake)
+      }
+  where
+    splitByPct :: Rational -> Rational -> Rational -> [a] -> ([a], [a], [a], [a])
+    splitByPct x y z l =
+      let
+        size = fromIntegral $ length l
+        (xs, rest) = splitAt (ceiling (x * size)) l
+        (ys, rest') = splitAt (ceiling (y * size)) rest
+        (zs, rest''') = splitAt (ceiling (z * size)) rest'
+       in
+        (xs, ys, zs, rest''')
+
+genRatios :: Gen Ratios
+genRatios = do
+  (a, b, c, d, e, _) <- genPctsOf100
+  pure $ Ratios {yes = a % 100, no = b % 100, abstain = c % 100, alwaysAbstain = d % 100, noConfidence = e % 100}
+
+genPctsOf100 :: Gen (Integer, Integer, Integer, Integer, Integer, Integer)
+genPctsOf100 = do
+  a <- choose (0, 100)
+  b <- choose (0, 100 - a)
+  c <- choose (0, 100 - b - a)
+  d <- choose (0, 100 - c - b - a)
+  e <- choose (0, 100 - d - c - b - a)
+  let f = 100 - e - d - c - b - a
+  pure (a, b, c, d, e, f)
```

Comment:
This is sort of number generation will lead to `f` and ` e` to be close to zero.
In other words, the only uniformly distributed value will be `a`, others will be biased.

It is a bit hard to generate uniformly distributed integral values that sum up to a number, but, if I understand the intent here correctly, that is not what you need. You need a number of ratios that sum up to 1:

```suggestion
genPcts :: Gen (Ratio, Ratio, Ratio, Ratio, Ratio, Ratio)
genPcts = do
  a <- choose (0, 100)
  b <- choose (0, 100)
  c <- choose (0, 100)
  d <- choose (0, 100)
  e <- choose (0, 100)
  f <- choose (0, 100)
  let s = a + b + c + d + e + f
  pure (a % s, b % s, c % s, d % s, e % s, f % s)
```

Reply:
:exploding_head:  This is so good! Thank you